DESIGN.rst

```diff
@@ -0,0 +1,53 @@
+============
+Design Notes
+============
+
+Introduction
+============
+
+TODO.
+
+Interfaces
+==========
+
+Rasterio has interfaces that are not yet described using abstract base classes
+or other formal interface system.
+
+DataAccessor
+------------
+
+This interface is involved with opening a dataset for access and is implemented
+by the DatasetReader and DatasetWriter classes. Their constructors take a str
+or os.PathLike object and, internally, attempt to adapt it to a
+rasterio.path.Path object.
+
+rasterio.path.Path
+------------------
+
+GDAL's GDALOpenEx takes an array of utf-8 encoded bytes as its primary
+argument. These bytes may contain a filename, a URL, an RDBMS connection
+string, XML, or JSON. Almost any kind of dataset address, really. GDAL puts no
+constraint on the content at all. A future format driver might use an array of
+emoji to address data and GDAL would be fine with that.
```

Comment:
I think we can remove the emoji mention?

Reply:
I'm going to leave it. It's anarchy in GDAL and new developers need to know.